File: CS50X_TS/PSET6/sentimental-credit/credit.py
from cs50 import get_string

# Prompt user for valid card number
while True:
    card_number = get_string("Number: ")
    if card_number.isnumeric():
        break

# Initialize sums for Luhn’s Algorithm
sum_even_positions = 0
sum_odd_positions = 0

# Reverse the card number and convert to list of integers
reversed_digits = reversed([int(digit) for digit in card_number])

# Iterate over the reversed digits using Luhn's Algorithm
for index, digit in enumerate(reversed_digits):
    if (index + 1) % 2 == 0:
        doubled_digit = digit * 2
        if doubled_digit > 9:
            sum_odd_positions += int(doubled_digit / 10) + (doubled_digit % 10)
        else:
            sum_odd_positions += doubled_digit
    else:
        sum_even_positions += digit

# Calculate the final checksum
final_sum = sum_even_positions + sum_odd_positions

# Extract card's first two digits and its length
first_two_digits = int(card_number[:2])
card_length = len(card_number)

# Check if the card passes Luhn’s Algorithm
if final_sum % 10 == 0:
    # Check for card types
    if first_two_digits in [34, 37] and card_length == 15:
        print("AMEX")
    elif (first_two_digits in range(51, 56) and card_length == 16):
        print("MASTERCARD")
    elif (int(card_number[0]) == 4 and card_length in [13, 16]):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: CS50X_TS/PSET1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool check_luhn(long number);
string get_card_type(long number);

int main(void)
{
    // Get the card number from the user
    long cardNumber = get_long("Number: ");

    // Check if the card number is valid using Luhn's algorithm
    if (check_luhn(cardNumber))
    {
        // Get the type of card based on the number
        string type = get_card_type(cardNumber);
        printf("%s\n", type);
    }
    else
    {
        printf("INVALID\n");
    }
}

bool check_luhn(long number)
{
    int sum = 0;
    bool alternate = false;

    while (number > 0)
    {
        int digit = number % 10;
        if (alternate)
        {
            digit *= 2;
            if (digit > 9)
                digit -= 9;
        }
        sum += digit;
        alternate = !alternate;
        number /= 10;
    }

    return (sum % 10) == 0;
}

string get_card_type(long number)
{
    int length = 0;
    long start_digits = number;

    // Find the length of the number and the starting digits
    while (start_digits >= 100)
    {
        start_digits /= 10;
        length++;
    }
    length += 2; // adjust length for the last two digits

    // Check the card type based on length and starting digits
    if ((start_digits == 34 || start_digits == 37) && length == 15)
    {
        return "AMEX";
    }
    else if (start_digits >= 51 && start_digits <= 55 && length == 16)
    {
        return "MASTERCARD";
    }
    else if ((start_digits / 10 == 4) && (length == 13 || length == 16))
    {
        return "VISA";
    }

    return "INVALID";
}
